array.c

```diff
@@ -2991,22 +2991,46 @@ rb_ary_join(VALUE ary, VALUE sep)
 
 /*
  *  call-seq:
- *     ary.join(separator=$,)    -> str
+ *    array.join ->new_string
+ *    array.join(separator = nil) -> new_string
  *
- *  Returns a string created by converting each element of the array to
- *  a string, separated by the given +separator+.
- *  If the +separator+ is +nil+, it uses current <code>$,</code>.
- *  If both the +separator+ and <code>$,</code> are +nil+,
- *  it uses an empty string.
+ *  Returns the new \String formed by joining the array elements after conversion.
+ *  For each element +element+
+ *  - Uses <tt>element.to_s</tt> if +element+ is not a <tt>kind_of?(Array)</tt>.
+ *  - Uses recursive <tt>element.join(separator)</tt> if +element+ is a <tt>kind_of?(Array)</tt>.
  *
- *     [ "a", "b", "c" ].join        #=> "abc"
- *     [ "a", "b", "c" ].join("-")   #=> "a-b-c"
+ *  Argument +separator+, if given, must be a
+ *  {String-convertible object}[doc/implicit_conversion_rdoc.html#label-String-Convertible+Objects].
  *
- *  For nested arrays, join is applied recursively:
+ *  ---
+ *
+ *  With no argument, joins using the output field separator, <tt>$,</tt>:
+ *    a = [:foo, 'bar', 2]
+ *    $, # => nil
+ *    a.join # => "foobar2"
+ *
+ *  With argument +separator+, joins using that separator:
+ *    a = [:foo, 'bar', 2]
+ *    a.join("\n") # => "foo\nbar\n2"
```

Comment:
This is yet another gotcha with `join` that may be worthy to document: the single or double quotes matter.
With double quotes:

```ruby
puts a.join("\n")
foo
bar
2
=> nil
```
With single quotes:
```ruby
 puts a.join('\n')
foo\nbar\n2
```

This is because the nature of the string escape sequences/characters in ruby (which, BTW, I cannot see a _clear_ reference in the documentation)

Reply:
We have some of this covered over in [doc/syntax/literals.rdoc](https://github.com/ruby/ruby/blob/master/doc/syntax/literals.rdoc#strings-), but if that's not sufficient we should be covering `""` vs `''` there or in String, not here

Reply:
Agree w/@drbrain.